### PR TITLE
Fields: include and exclude empty values

### PR DESCRIPTION
**Pull request by gtk-grafana**
Fixes: https://github.com/grafana/explore-logs/issues/590

Updates the buttons on the field names view to allow users to include/exclude all logs that contain that field.

Base branch: https://github.com/grafana/explore-logs/pull/682
Should not be merged until parent branch has been merged!

Fields:
image

Labels:
image

Tooltips:
image

image

image
image
image

### CODE REVIEW

**Comment by gtk-grafana** [comment]
Opened up for review after implementing the above, and reviewing UI with @zizzpudding.

**Comment by matyax** [comment]
I'm seeing the following issue:

https://github.com/user-attachments/assets/84fd512e-c8c4-49fd-84b6-d7e42cb3afa9

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx [hunk 1]
```diff
@@ -1,24 +1,137 @@
...
+    });
+
     return (
-      <Button variant="secondary" size="sm" onClick={model.onClick} aria-label={`Select ${model.useState().labelName}`}>
-        Select
-      </Button>
+      <>
+        {showFilterField === true && (
+          <FilterButton
+            isExcluded={existingFilter?.operator === '='}
```
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:36]: May want to use `FilterOp`.

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx [hunk 2]
```diff
@@ -1,24 +1,137 @@
...
+      if (logLinesWithLabelCount < logsPanelData.length) {
+        this.setState({
+          showFilterField: true,
+        });
+      } else {
+        this.setState({
+          showFilterField: false,
+        });
+      }
+    }
```
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:135]: Nit:

```suggestion
    if (!labels || !logsPanelData) {
      this.setState({
        showFilterField: false,
      });
      return;
    }
    // iterate through all the labels on the log panel query result and count how many times this exists
    const logLinesWithLabelCount = labels.values.reduce((acc, labels) => {
      if (labels?.[this.state.labelName]) {
        acc++;
      }
      return acc;
    }, 0);

    const panel = sceneGraph.getAncestor(this, VizPanel);
    const percentage = ((logLinesWithLabelCount / logsPanelData.length) * 100).toLocaleString();
    const description = `${this.state.labelName} exists on ${percentage}% of ${logsPanelData.length} sampled log lines`;

    // Update the desc
    panel.setState({
      description,
    });

    if (logLinesWithLabelCount < logsPanelData.length) {
      this.setState({
        showFilterField: true,
      });
    }
```

File: src/services/query.ts [hunk 3]
```diff
@@ -66,6 +66,9 @@ export function renderLogQLFieldFilters(filters: AdHocVariableFilter[]) {
 }
 
 function renderFilter(filter: AdHocVariableFilter) {
+  if (filter.value === '""') {
```
Comment (matyax) [inline src/services/query.ts:69]: Would it break everything if the value is literally an empty string? I dislike seeing the "" in the field.

**Review by matyax — COMMENTED** [review]
Few comments, but looks great.

File: src/services/query.ts (hunk 3 above)
Comment (gtk-grafana) [reply src/services/query.ts:98]: Probably, i'll take a look tho

File: src/services/variables.ts [hunk 4]
```diff
@@ -27,6 +27,8 @@ export const ALL_VARIABLE_VALUE = '$__all';
 export const LEVEL_VARIABLE_VALUE = 'detected_level';
 export const PATTERNS_TEXT_FILTER = 'patternsFilter';
 
+export const EmptyVariableValue = '""';
```
Comment (matyax) [inline src/services/variables.ts:30]:
```suggestion
export const EMPTY_VARIABLE_VALUE = '""';
```

👉 👈

Comment (gtk-grafana) [reply src/services/variables.ts]: Done!

File: src/services/query.ts (hunk 3 above)
Comment (gtk-grafana) [reply src/services/query.ts]: So yeah, empty string means no filter in the variable, and is evaluated falsey in a bunch of places. This is the correct value for the query, I think we'd need more control over adhoc variables to get this to work

Comment (matyax) [reply src/services/query.ts]: I imagined that was a possibility. Thanks for checking!

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx [hunk 5]
```diff
@@ -1,24 +1,177 @@
...
+  getLabelsVariable,
+  getLevelsVariable,
+  LEVEL_VARIABLE_VALUE,
+  SERVICE_NAME,
+} from '../../../services/variables';
+import { AdHocVariableFilter, Field, Labels, LoadingState } from '@grafana/data';
+import { FilterOp } from '../../../services/filters';
 
-interface SelectLabelActionSceneState extends SceneObjectState {
+interface SelectFieldActionSceneState extends SceneObjectState {
```
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:27]:
```suggestion
interface SelectLabelActionSceneState extends SceneObjectState {
```

File: src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx [hunk 6]
```diff
@@ -1,24 +1,177 @@
...
+    const variable = model.getVariable();
+    const variableName = variable.useState().name;
+    const existingFilter = model.getExistingFilter(variable);
+
     return (
-      <Button variant="secondary" size="sm" onClick={model.onClick} aria-label={`Select ${model.useState().labelName}`}>
-        Select
-      </Button>
+      <>
+        {variable && showFilterField === true && (
```
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/SelectLabelActionScene.tsx:48]: If variable could not be defined, then `const variableName = variable.useState().name;` (line 43) should throw. So either one or the other.

File: src/services/query.ts [hunk 7]
```diff
@@ -95,6 +95,9 @@ export function renderLogQLFieldFilters(filters: AdHocVariableFilter[]) {
 }
 
 function renderFilter(filter: AdHocVariableFilter) {
+  if (filter.value === EMPTY_VARIABLE_VALUE) {
```
Comment (matyax) [inline src/services/query.ts:98]: Silly, but took me a minute to understand this. Maybe adding a comment would be nice.

**Review by matyax — COMMENTED** [review]
Left a few minor comments but looks ready do go 👌

**Review by matyax — APPROVED** [review]
Feel free to follow up on the comments and merge 🫡

File: src/services/query.ts (hunk 7 above)
Comment (gtk-grafana) [reply src/services/query.ts:99]: Done!

**Pull request merged**